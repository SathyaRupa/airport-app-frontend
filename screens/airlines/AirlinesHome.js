import {
  StyleSheet,
  FlatList,
  ActivityIndicator,
  View,
  SafeAreaView,
} from 'react-native';
import ItemCard from '../../components/ItemCard';
import * as React from 'react';
import {useEffect, useState} from 'react';
import airlinesService from '../../helpers/airlinesService';
import {SuccessToast, ErrorToast} from '../../components/ToastMessage';
import {
  Avatar,
  PaperProvider,
  Portal,
  Text,
  Modal,
  Button,
} from 'react-native-paper';
import CreateButton from '../../components/CreateButton';
import {useIsFocused} from '@react-navigation/native';
import Toast from 'react-native-toast-message';

function AirlinesHome({navigation}) {
  const [airlines, setAirlines] = useState([]);
  const [page, setPage] = useState(0);
  const [loading, setLoading] = useState(true);
  const [allDataLoaded, setAllDataLoaded] = useState(false);
  const isFocused = useIsFocused();
  const [visible, setVisible] = useState(false);
  const [selectedAirline, setSelectedAirline] = useState({
    id: null,
    name: '',
  });

  const showModal = () => setVisible(true);
  const hideModal = () => setVisible(false);

  useEffect(() => {
    async function fetchData() {
      if (!allDataLoaded && isFocused) {
        const response = await airlinesService.fetchAll(page);
        if (response.length === 0) {
          setAllDataLoaded(true);
          return;
        }
        setAirlines(prevAirlines => [...prevAirlines, ...response]);
      }
      setLoading(false);
    }
    fetchData();
  }, [page, allDataLoaded, isFocused]);

  const loadMoreAirlines = () => {
    if (!allDataLoaded) {
      setPage(prevPage => prevPage + 1);
    }
  };

  useEffect(() => {
    if (isFocused) {
      setAirlines([]);
      setPage(0);
      setAllDataLoaded(false);
      setLoading(true);
    }
  }, [isFocused]);

  const renderFooter = () => {
    return loading ? <ActivityIndicator size="large" color="#4D869C" /> : null;
  };

  const handlePress = id => {
    navigation.push('Airline Details', {id});
  };
  const handleUpdate = (id, name) => {
    setSelectedAirline({id, name});
    setAllDataLoaded(false);
    setAirlines([]);
    setPage(0);
    navigation.push('Update Airline', {id});
  };

  const handleDelete = (id, name) => {
    setSelectedAirline({id, name});
    showModal();
  };

  const deleteAirlineById = async () => {
    try {
      const response = await airlinesService.deleteAirline(selectedAirline.id);
      if (response.status === 200) {
        SuccessToast(response.data);
      } else {
        ErrorToast(response.data);
      }
      setAirlines(prevAirlines =>
        prevAirlines.filter(airline => airline.id !== selectedAirline.id),
      );
      hideModal();
    } catch (error) {
      ErrorToast();
      console.error('Error deleting airline:', error);
    }
  };

  const icon = (
    <Avatar.Image
      style={{backgroundColor: 'transparent'}}
      size={50}
      source={require('../../assets/icons/airlines.png')}
    />
  );

  return (
    <PaperProvider>
      <SafeAreaView style={styles.mainContainer}>
        <CreateButton
          handleOnPress={() => {
            setAllDataLoaded(false);
            setAirlines([]);
            setPage(0);
            navigation.push('Create Airline');
          }}
          testId="create-button"
        />

        <FlatList
          data={airlines}
          renderItem={itemData => (
            <ItemCard
              id={itemData.item.id}
              name={itemData.item.name}
              icon={icon}
              onPress={() => handlePress(itemData.item.id)}
              handleDelete={handleDelete}
              handleUpdate={handleUpdate}
              testId="item-card"
            />
          )}
          onEndReached={loadMoreAirlines}
          onEndReachedThreshold={0.5}
          ListFooterComponent={renderFooter}
        />

        <Portal>
          <Modal
            visible={visible}
            onDismiss={hideModal}
            contentContainerStyle={styles.modalContent}>
            <View>
              <Text style={styles.modalText}>
                Are you sure you want to delete the airline "
                {selectedAirline.name}"?
              </Text>
              <View style={styles.modalButtons}>
                <Button
                  style={styles.button}
                  mode="contained"
                  onPress={deleteAirlineById}>
                  Yes
                </Button>
                <Button
                  style={styles.button}
                  mode="contained"
                  onPress={hideModal}>
                  No
                </Button>
              </View>
            </View>
          </Modal>
        </Portal>
        <Toast />
      </SafeAreaView>
    </PaperProvider>
  );
}

const styles = StyleSheet.create({
  mainContainer: {
    padding: 10,
    flex: 1,
  },
  modalContent: {
    backgroundColor: 'white',
    padding: 20,
    marginHorizontal: 20,
    borderRadius: 10,
  },
  modalText: {
    fontSize: 20,
    color: 'black',
    textAlign: 'center',
    marginBottom: 20,
  },
  modalButtons: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
  },
  button: {
    margin: 10,
    backgroundColor: '#7AB2B2',
    width: 100,
    height: 40,
  },
});

export default AirlinesHome;
