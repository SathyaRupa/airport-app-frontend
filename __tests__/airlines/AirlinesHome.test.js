import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import AirlinesHome from '../../screens/airlines/AirlinesHome';
import AirlineService from '../../helpers/AirlineService';
import {useIsFocused} from '@react-navigation/native';

jest.mock('react-native-linear-gradient', () => 'LinearGradient');
jest.mock('@react-navigation/native', () => ({
  useIsFocused: jest.fn(),
}));
jest.mock('../../helpers/AirlineService', () => ({
  fetchAll: jest.fn(),
  delete: jest.fn(),
}));

jest.mock('react-native-toast-message', () => 'Toast');
jest.mock('../../components/ToastMessage', () => ({
  SuccessToast: jest.fn(),
  ErrorToast: jest.fn(),
}));

describe('Airlines home - get all airlines', () => {
  let mockNavigation;

  beforeEach(() => {
    mockNavigation = {
      push: jest.fn(),
    };
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  it('snapshot test', () => {
    const airlinesHome = render(<AirlinesHome navigation={mockNavigation} />);

    expect(airlinesHome).toMatchSnapshot();
  });

  it('should render an item card', async () => {
    useIsFocused.mockReturnValue(true);

    AirlineService.fetchAll.mockResolvedValueOnce([
      {name: 'Jet Airways', count: '6'},
    ]);
    const airlinesHome = render(<AirlinesHome navigation={mockNavigation} />);
    await waitFor(() => {
      expect(airlinesHome.queryByTestId('item-card-0')).toBeTruthy();
      expect(AirlineService.fetchAll).toHaveBeenCalledWith(0);
      expect(airlinesHome.getByText('Jet Airways')).toBeTruthy();
    });
  });

  it('should render an item card and naviage to update page when update icon is pressed', async () => {
    useIsFocused.mockReturnValue(true);

    AirlineService.fetchAll.mockResolvedValueOnce([
      {id: '1', name: 'Jet Airways', count: '6'},
      {id: '2', name: 'Kingfisher Airways', count: '7'},
    ]);

    const {getAllByTestId} = render(
      <AirlinesHome navigation={mockNavigation} />,
    );

    await waitFor(() => {
      const updateButton = getAllByTestId('update-icon');
      expect(updateButton[0]).toBeTruthy();
      expect(mockNavigation.push).not.toHaveBeenCalled();
      fireEvent.press(updateButton[0]);
    });
    expect(mockNavigation.push).toHaveBeenCalledWith('Update Airline', {
      id: '1',
    });
  });

  it('should render an item card and a model to pop up when delete icon is clicked', async () => {
    useIsFocused.mockReturnValue(true);

    AirlineService.fetchAll.mockResolvedValueOnce([
      {id: '1', name: 'Jet Airways', count: '6'},
      {id: '2', name: 'Kingfisher Airways', count: '7'},
    ]);

    const {getAllByTestId, getByText} = render(
      <AirlinesHome navigation={mockNavigation} />,
    );

    await waitFor(() => {
      const deleteButton = getAllByTestId('delete-icon');
      expect(deleteButton[0]).toBeTruthy();
      expect(mockNavigation.push).not.toHaveBeenCalled();
      fireEvent.press(deleteButton[0]);
    });
    expect(
      getByText('Are you sure you want to delete the airline\n "Jet Airways"?'),
    ).toBeTruthy();
  });

  it('should render an item card and delete the card when confirmed from the model', async () => {
    useIsFocused.mockReturnValue(true);

    AirlineService.fetchAll.mockResolvedValueOnce([
      {id: '1', name: 'Jet Airways', count: '6'},
      {id: '2', name: 'Kingfisher Airways', count: '7'},
    ]);
    AirlineService.delete.mockResolvedValueOnce({
      data: 'Deleted the airline successfully',
      status: 200,
    });

    const {getAllByTestId, getByText, getByTestId} = render(
      <AirlinesHome navigation={mockNavigation} />,
    );

    await waitFor(() => {
      const deleteButton = getAllByTestId('delete-icon');
      expect(deleteButton[0]).toBeTruthy();
      expect(mockNavigation.push).not.toHaveBeenCalled();
      fireEvent.press(deleteButton[0]);
    });
    expect(
      getByText('Are you sure you want to delete the airline\n "Jet Airways"?'),
    ).toBeTruthy();
    fireEvent.press(getByTestId('yes-button'));
    expect(AirlineService.delete).toHaveBeenCalledWith('1');
  });

  it('should not render an item card when no data is recieved', async () => {
    AirlineService.fetchAll.mockResolvedValue([]);

    const airlinesHome = render(<AirlinesHome navigation={mockNavigation} />);
    expect(airlinesHome.queryByTestId('item-card-0')).toBeNull();
    await waitFor(() =>
      expect(airlinesHome.queryByText('Jet Airways')).toBeNull(),
    );
  });

  it('should navigate to Create Airlines page when clicked on create', () => {
    const {getByText} = render(<AirlinesHome navigation={mockNavigation} />);
    const createButton = getByText('Create');
    expect(createButton).toBeTruthy();
    fireEvent.press(createButton);
    expect(mockNavigation.push).toHaveBeenCalledWith('Create Airline');
  });
});
