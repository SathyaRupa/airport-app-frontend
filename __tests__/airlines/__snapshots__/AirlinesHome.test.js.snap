// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Airlines home - get all airlines snapshot test 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <RCTSafeAreaView
      style={
        {
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row-reverse",
            "padding": 10,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#7AB2B2",
              "borderRadius": 20,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": 120,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#7AB2B2",
                "borderColor": "transparent",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 0,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Create
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <RCTScrollView
        ListFooterComponent={[Function]}
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0.5}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onLayout={[Function]}
          >
            <ActivityIndicator
              color="#4D869C"
              size="large"
            />
          </View>
        </View>
      </RCTScrollView>
      <Toast />
    </RCTSafeAreaView>
  </View>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</RNCSafeAreaProvider>
`;
